Add creation of XDG directories

nydragon opened the pull request:
This PR aims to add the creation of XDG directories, for use by direct users of hjem but also downstream tools such as hjem-rum.

nydragon commented on modules/nixos/default.nix:
```diff
@@ -88,22 +88,35 @@ in {
+                # is constructed.
+                "${mode} '${file.target}' - - - - ${file.source}"
+            ) (filter (f: f.enable && f.source != null) (attrValues files))
+            ++ (optionals xdg.enable [
+              "d '${xdg.dataDirectory}'"
+              "d '${xdg.stateDirectory}'"
+              "d '${xdg.cacheDirectory}'"
+              "d '${xdg.configDirectory}'"
+            ]);
```
Should this potentially be in its own file inside of the nixos folder ? so that each module has a matching implementation.

fxzzi commented:
tested, impl seems to work perfectly for me. Of course I have no say in the structure of the module tho. :)

nydragon commented:
That would definitely be more sustainable, do you have a preference as to how ?

I could imagine extending `hjem.files.type` in a way that allows setting it to be a directory (although that would make options more confusing, as it makes no sense to set executable/text/source if the file is a directory)

nydragon commented:
Updated to add a `type` option to the files attrs. Default is symlink but can be set to directory if one should be created instead. I also moved the commits from #32 as I needed them.

nydragon commented:
> rather than just creating the directories I'd rather expose a way to create symlinks in those dirs.

This is not always desired behaviour, user directories like Pictures, Desktop, etc. should exist, without the need to symlink something into them.

If you'd prefer to upstream your approach I will go ahead and close this PR

llakala commented on modules/common.nix:
```diff
@@ -195,15 +207,23 @@ in {
+        }
+      ]
+      ++ (mapAttrsToList (
+          _target: file: {
```
```suggestion
          _: file: {
```
If you're not using a variable, I don't see the point of keeping its name.

nydragon replied:
Some people prefer to keep the name for legibility, have no strong opinion on it. Will update it once I know whether we go forward with this PR

llakala replied:
Sounds good.

eclairevoyant replied:
I have no opinion, either is fine, the underscore-prefix is a soft indicator that it's unused (and a convention that is shared in nixpkgs' lib).

llakala reviewed (commented):
Tiny thing I saw while reading. Rest is readable enough for my tastes.

eclairevoyant commented:
> > rather than just creating the directories I'd rather expose a way to create symlinks in those dirs.
> 
> This is not always desired behaviour, user directories like Pictures, Desktop, etc. should exist, without the need to symlink something into them.
> 
> If you'd prefer to upstream your approach I will go ahead and close this PR

I get what you're saying but I don't know of any xdg spec that handles pics/desktop/etc. and I believe most programs would have to handle the case that XDG_whatever_HOME doesn't exist. But I guess it doesn't hurt to have these options as well as have clearly defined values for the xdg dirs, I'll try to make my future changes work with this PR.

eclairevoyant commented:
closing as we went with https://github.com/feel-co/hjem/pull/45
